### PR TITLE
[FEATURE][processing] Add a save features to file algorithm 

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR adds a new native processing algorithm to save vector features to a file dataset. It comes with a pair of advanced features to pass on GDAL dataset and layer options.

It comes in handy as part of models. E.g.:
Screenshot from 2020-07-21 15-04-07

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+QString QgsSaveFeaturesAlgorithm::displayName() const
+{
+  return QObject::tr( "Save vector features to file" );
+}
+
+QStringList QgsSaveFeaturesAlgorithm::tags() const
+{
+  return QObject::tr( "save,write" ).split( ',' );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:34]:
```suggestion
  return QObject::tr( "save,write,export" ).split( ',' );
```

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+QString QgsSaveFeaturesAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeneral" );
+}
+
+QString QgsSaveFeaturesAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Save vector features to a specified file dataset." );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:49]:
```suggestion
  return QObject::tr( "This algorithm saves vector features to a specified file dataset." );
```

+ some description about the dataset/layer options, and where to find help for them

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+QgsSaveFeaturesAlgorithm *QgsSaveFeaturesAlgorithm::createInstance() const
+{
+  return new QgsSaveFeaturesAlgorithm();
+}
+
+void QgsSaveFeaturesAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Vector features" ) ) );
+  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Saved" ), QObject::tr( "All files (*.*)" ), QVariant(), false ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:60]:
```suggestion
  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Saved" ), QgsVectorFileWriter::fileFilterString(), QVariant(), false ) );
```

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+
+  QString layerName = parameterAsString( parameters, QStringLiteral( "LAYER_NAME" ), context ).trimmed();
+  QVariantMap createOptions;
+  if ( !layerName.isEmpty() )
+  {
+    createOptions[QStringLiteral( "layerName" )] = layerName;
+  }
+
+  QStringList datasourceOptions = parameterAsString( parameters, QStringLiteral( "DATASOURCE_OPTIONS" ), context ).trimmed().split( QLatin1Char( ';' ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:84]:
```suggestion
  QStringList datasourceOptions = parameterAsString( parameters, QStringLiteral( "DATASOURCE_OPTIONS" ), context ).trimmed().split( ';', Qt::SkipEmptyParts );
```

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  QStringList layerOptions = parameterAsString( parameters, QStringLiteral( "LAYER_OPTIONS" ), context ).trimmed().split( QLatin1Char( ';' ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:85]:
```suggestion
  QStringList layerOptions = parameterAsString( parameters, QStringLiteral( "LAYER_OPTIONS" ), context ).trimmed().split( ';', Qt::SkipEmptyParts );
```

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, source->fields(),
+                                          source->wkbType(), source->sourceCrs(), QgsFeatureSink::SinkFlags(), createOptions, datasourceOptions, layerOptions ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  double step = source->featureCount() > 0 ? 50.0 / source->featureCount() : 1;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:94]:
```suggestion
  double step = source->featureCount() > 0 ? 100.0 / source->featureCount() : 1;
```

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  int i = 0;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:95]:
```suggestion
  long long i = 0;
```

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+  QgsFeatureIterator features = source->getFeatures( QgsFeatureRequest(), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+  QgsFeature feat;
+
+  QHash< QVariant, QList< double > > means;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:100]:
```suggestion
```

File: src/core/processing/qgsprocessingalgorithm.h [hunk 9]
```diff
@@ -684,7 +684,7 @@ class CORE_EXPORT QgsProcessingAlgorithm
      * \throws QgsProcessingException
      */
     QgsFeatureSink *parameterAsSink( const QVariantMap &parameters, const QString &name, QgsProcessingContext &context, QString &destinationIdentifier SIP_OUT,
-                                     const QgsFields &fields, QgsWkbTypes::Type geometryType = QgsWkbTypes::NoGeometry, const QgsCoordinateReferenceSystem &crs = QgsCoordinateReferenceSystem(), QgsFeatureSink::SinkFlags sinkFlags = QgsFeatureSink::SinkFlags() ) const SIP_THROW( QgsProcessingException ) SIP_FACTORY;
+                                     const QgsFields &fields, QgsWkbTypes::Type geometryType = QgsWkbTypes::NoGeometry, const QgsCoordinateReferenceSystem &crs = QgsCoordinateReferenceSystem(), QgsFeatureSink::SinkFlags sinkFlags = QgsFeatureSink::SinkFlags(), const QVariantMap &createOptions = QVariantMap(), const QStringList &datasourceOptions = QStringList(), const QStringList &layerOptions = QStringList() ) const SIP_THROW( QgsProcessingException ) SIP_FACTORY;
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingalgorithm.h:687]: Can you add docs for the new arguments?

File: src/core/processing/qgsprocessingparameters.h [hunk 10]
```diff
@@ -937,7 +937,7 @@ class CORE_EXPORT QgsProcessingParameters
      */
     static QgsFeatureSink *parameterAsSink( const QgsProcessingParameterDefinition *definition, const QVariantMap &parameters,
                                             const QgsFields &fields, QgsWkbTypes::Type geometryType, const QgsCoordinateReferenceSystem &crs,
-                                            QgsProcessingContext &context, QString &destinationIdentifier SIP_OUT, QgsFeatureSink::SinkFlags sinkFlags = QgsFeatureSink::SinkFlags() ) SIP_FACTORY;
+                                            QgsProcessingContext &context, QString &destinationIdentifier SIP_OUT, QgsFeatureSink::SinkFlags sinkFlags = QgsFeatureSink::SinkFlags(), const QVariantMap &createOptions = QVariantMap(), const QStringList &datasourceOptions = QStringList(), const QStringList &layerOptions = QStringList() ) SIP_FACTORY;
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.h:940]: Can you add docs for the new arguments?

File: src/core/processing/qgsprocessingutils.cpp [hunk 11]
```diff
@@ -737,11 +738,13 @@ QgsFeatureSink *QgsProcessingUtils::createFeatureSink( QString &destination, Qgs
       // use QgsVectorFileWriter for OGR destinations instead of QgsVectorLayerImport, as that allows
       // us to use any OGR format which supports feature addition
       QString finalFileName;
+      QString finalLayerName;
       QgsVectorFileWriter::SaveVectorOptions saveOptions;
       saveOptions.fileEncoding = options.value( QStringLiteral( "fileEncoding" ) ).toString();
+      saveOptions.layerName = !layerName.isEmpty() ? layerName : options.value( QStringLiteral( "layerName" ) ).toString();
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingutils.cpp:744]: Can you add a test to testqgsprocessing.cpp checking the behavior of the layerName handling?

File: tests/src/analysis/testqgsprocessingalgs.cpp [hunk 12]
```diff
@@ -235,6 +236,32 @@ QVariantMap pkgAlg( const QStringList &layers, const QString &outputGpkg, bool o
...
+  const QgsProcessingAlgorithm *saveFeatures( QgsApplication::processingRegistry()->algorithmById( QStringLiteral( "native:savefeatures" ) ) );
+  std::unique_ptr< QgsProcessingContext > context = qgis::make_unique< QgsProcessingContext >();
+  context->setProject( QgsProject::instance() );
+
+  QgsProcessingFeedback feedback;
+  bool ok = false;
+  QVariantMap outputs = saveFeatures->run( parameters, *context, &feedback, &ok );
+  QCOMPARE( ok, true );
+  QCOMPARE( outputs.value( QStringLiteral( "OUTPUT" ) ).toString(), outputGeoJson );
+}
```
Comment (nyalldawson) [inline tests/src/analysis/testqgsprocessingalgs.cpp:263]: Better also read back in the created file and confirm that the precision argument has been correctly applied

**Review by nyalldawson — COMMENTED** [review]
Great addition -- thanks!

**Comment by nirvn** [comment]
@nyalldawson , comments addressed, thanks for the review. Travis is green, I'll merge when you give me a thumbs up.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**